File: leetcode/algorithms/455_assign_cookies/main.py
from typing import List


class Solution:
    def findContentChildren(self, g: List[int], s: List[int]) -> int:
        g = sorted(g)
        s = sorted(s)
        gi = 0
        for sv in s:
            if gi >= len(g):
                break
            if sv >= g[gi]:
                gi += 1
        return gi


    # Best Solution
    def bestSolution(self, g: List[int], s: List[int]) -> int:
        g.sort()
        s.sort()

        childi = 0
        cookiei = 0

        while cookiei < len(s) and childi < len(g):
            if s[cookiei] >= g[childi]:
                childi += 1
            cookiei += 1

        return childi

File: leetcode/algorithms/455_assign_cookies/main.cpp
#include <vector>
using namespace std;

class AssignCookies {
public:
    int findContentChildren(vector<int> &g, vector<int> &s) {
        sort(g.begin(), g.end());
        sort(s.begin(), s.end());

        int i = 0;

        for (int j = 0; i < g.size() && j < s.size(); j++) {
            if (g[i] <= s[j]) {
                i++;
            }
        }

        return i;
    }

    // Best Solution
    int bestSolution(vector<int> &g, vector<int> &s) {
        int cookiesNums = s.size();

        if (cookiesNums == 0) {
            return 0;
        }

        sort(g.begin(), g.end());
        sort(s.begin(), s.end());

        int maxNum = 0;
        int cookieIndex = cookiesNums - 1;
        int childIndex = g.size() - 1;

        while (cookieIndex >= 0 && childIndex >= 0) {
            if (s[cookieIndex] >= g[childIndex]) {
                maxNum++;
                cookieIndex--;
                childIndex--;
            } else {
                childIndex--;
            }
        }

        return maxNum;
    }
};
